### DISCUSSION TITLE
Set Allow List and Matter Reference updates

### DISCUSSION THREAD

**Discussion by hagen31** [sadieblue/agilityblue.workspace-scripts#19] [Q&A] [answered]
I have two questions about how I can enhance my new matter creation script.

1. I have it set it where if a Client Prefix matches a specific value, lets's say ABCD it will set the AccessStateID to 5 and AccessStateName to Allow List. I've been unable to find anything related to taking that a step further and setting the names on the Allow List is that possible?

2. I have it set so the Matter Reference field only accepts 5 digits and then will prepend the Client Prefix + _ to what the user enters there after saving. I have another question about automatically pulling in the next available numeric value from a MediaLogEntry reference and similar concept here I want the system to automatically find the next available numeric value (across any client prefix this time) and automatically calculate the Matter Reference values. Example if there is an ABCD_24001 and EFGH_240002 the next new matter would be 240003 regardless of the associated Client Prefix. I've found several examples to identify the highest number on my Matters table; however, it is not pulling in all data and not getting me the next available number.

Does anybody have any thoughts or suggestions?

**Comment by bmward80** [comment]
Hi Austin,

For your first question, the allow/deny lists for matter access are still in feature preview and as a result, they have not been greenlit for scripting until the feature has been upgraded to stable and generally available.

For your second question, Can you provide your examples so I can identify why your data is not being pulled?

Thanks,
Brandon

**Reply by hagen31** [reply]
Hello Brandon,

Thank you for confirming the answer to question 1.

For question 2. I didn't save the code, lesson learned to save, that I put together to pull the Matter Reference for all Matters so will work to reconstruct that and update here when I can.

**Reply by hagen31** [reply]
The code works below if I use a Trigger of On Update After Save just to verify, I can get a numerical value to increase now trying to use this to understand how to expand across All Matters and unable to recreate my somewhat success from yesterday.

```powershell
# Get the matter object
$matter = Get-InboundObjectInstance

# Check if the matter object is valid
if ($null -eq $matter)
{
    Write-Output "Error: Matter object is null."
    return
}

# Check if the matter has a reference property
if (-not ($null -eq $matter.Reference -and $matter.Reference.Length -gt 0))
{
    Write-Output "Has matter reference: $($matter.Reference)"

    # Extract the part after the first underscore
    $referenceParts = $matter.Reference -split "_", 2
    if ($referenceParts.Length -eq 2)
    {
        $prefix = $referenceParts[0]
        $numberPart = $referenceParts[1]

        # Initialize the number
        $number = 0

        # Check if the number part is an integer
        if ([int]::TryParse($numberPart, [ref]$number))
        {
            # Increment the numeric part
            $number++

            # Format the incremented number with leading zeros
            $newNumber = "{0:D5}" -f $number

            # Combine the prefix and the incremented number
            $matter.Reference = "$($prefix)_$($newNumber)"
        }
        else
        {
            # If the number part is not an integer, handle the error or set a default value
            Write-Output "Invalid number part in matter reference. Starting with 00001."
            $matter.Reference = "$($prefix)_00001"
        }
    }
    else
    {
        # If there is no underscore or not enough parts, handle the error or set a default value
        Write-Output "Matter reference does not contain an underscore or is not in the expected format. Starting with 00001."
        $matter.Reference = "$($matter.Reference)_00001"
    }

    # Output the result
    Write-Output "Updated matter reference as $($matter.Reference)"
}
else
{
    Write-Output "Matter does not have a valid reference."
}
```
(edit to wrap code in a readable format)

**Reply by bmward80** [reply]
Hi Austin, I'll take a look at this and run it in my test environment.

Brandon

**Reply by hagen31** [reply]
Hello Brandon,

Thank you. Just to clarify the final trigger should be On Create Before Save so the Matter Reference field automatically takes the Reference from the Client table adds a _ then adds the next available number. The current code stayed on Matter Reference just to work through logic of getting a number stored.

**Reply by bmward80** [reply]
Hi Austin,

To get this script to work for a before save event, all you would need to do is add a return statement at the end of the script like the following:

```powershell
# previous section of the script omitted

else
{
    Write-Output "Matter does not have a valid reference."
}

return $matter
```

When creating a "before save" type of script, the system will expect that the same input type is also the same return type.

For your question about using the client reference to append to the matter reference, you could retrieve the client that the matter is associated with and append that to your matter reference. Use the `Get-Client` command to do that:

```powershell
# First, get the client...
$client = Get-Client -Id $matter.ClientId

# Then, get the reference for the client
$clientReference = $client.Reference
```

Matt showed you how you can filter for numbers that have a reference prefix using a `startswith()` filter and increment a number [in this post here](https://github.com/sadieblue/agilityblue.workspace-scripts/discussions/18#discussioncomment-10396382). You will just want to apply that filter to the [matters collection](https://help.agilityblue.com/docs/scripting-get-matters) using your client reference. That should get you pointed in the direction for what you're looking to do.

Brandon

**Reply by hagen31** [reply]
Hello Brandon,

Previously it wasn't getting the next available number. I'll see if I can recreate it.

**Reply by hagen31** [reply]
I was able to get back to the code where I went through the Matters collection to see what the next number was. Right now I have 377 matters and the highest number should be 24130 the code below says it's 22145. Could you test in your environment to see if you get the expected results?

```powershell
# Retrieve all Matter objects
$mattersResponse = Get-Matters

# Extract the actual collection of matter objects
$matters = $mattersResponse.Collection

# Initialize variable to track the highest number
$highestNumber = 0

# Process each Matter object
foreach ($matter in $matters) {

    # Check if the matter object has a reference property and that it is not empty
    if ($null -ne $matter['Reference'] -and $matter['Reference'].Length -gt 0) {
        # Extract the part after the first underscore
        $referenceParts = $matter['Reference'] -split "_", 2
        if ($referenceParts.Length -eq 2) {
            $numberPart = $referenceParts[1]  # Extract the number part (after the underscore)
            
            # Initialize the number
            $number = 0

            # Check if the number part is an integer
            if ([int]::TryParse($numberPart, [ref]$number)) {
                # Update the highest number if this number is greater
                if ($number -gt $highestNumber) {
                    $highestNumber = $number
                }
            } else {
                # Debug output if parsing fails
                Write-Output "Failed to parse number part: $($numberPart)"
            }
        } else {
            # Debug output if split fails
            Write-Output "Failed to split reference: $($matter['Reference'])"
        }
    } else {
        # Debug output if reference is null or empty
        Write-Output "Reference is null or empty for matter: $($matter)."
    }
}

# Increment the highest number found
$nextNumber = $highestNumber + 1

# Format the incremented number with leading zeros
$formattedNumber = "{0:D5}" -f $nextNumber  # Ensure 5 digits for the number

# Setup for new matter
$matter = Get-InboundObjectInstance

# Set the matter reference field
$matter.reference = "$($matter.ClientReference)_$($formattedNumber)"

Write-Output "The current highest number is $($highestNumber). The next highest number is $($formattedNumber)."
return $matter
```

**Reply by bmward80** [reply]
Running this in my environment won't really do anything since I don't have client references and matter references like you have setup. But one thing I notice, if you're trying to iterate through all of your matters, you are only pulling in the first 40 matters (the default for collections).

Check out this article on [how collections work](https://help.agilityblue.com/docs/scripting-tutorials-collections-basics). You will need to issue a higher TOP parameter so you can retrieve more records. You can request up to a maximum of 1,000 records, which is well within your range. Once you start needing more than that, you will need to work on a [paging strategy](https://help.agilityblue.com/docs/scripting-tutorials-collections-paging).

```powershell
$mattersResponse = Get-Matters -Top 1000
```

Brandon

**Reply by hagen31** [reply]
I needed a break from this script and was working on one to force Client Reference to unique and came across the same issue where it wasn't returning everything. That was easier script to work with and was just getting into the Top and paging strategy. I got paging to work there and will try it on this one.

Thank you for your guidance.

**Reply by hagen31** [reply, marked as answer]
This is complete with code as below. 
```powershell
# Initialize variable to track the highest number
$highestNumber = 0

# Initialize pagination variables
$top = 1000  # Number of records per page
$skip = 0    # Records to skip
$count = 0   # Total records processed
$allMatters = @() # Array to accumulate all client records

do {
    # Retrieve all Matter objects with pagination
    $mattersResponse = Get-Matters -Top $top -Skip $skip
    
    # Check if the response is valid and contains a collection
    if ($mattersResponse -and $mattersResponse.Collection) {
        $matters = $mattersResponse.Collection
        $allMatters += $matters

        # Update the count of processed records
        $count += $matters.Count

        # Calculate pagination details
        $page = ($skip / $top) + 1
        $totalPages = [Math]::Ceiling($mattersResponse.TotalCount / $top)

        Write-Output "Currently on page $page of $totalPages pages (encountered $($count) records at $top records per page)"

        # Update the skip value for the next page
        $skip += $top
    } else {
        break
    }
} while ($skip -lt $mattersResponse.TotalCount)

# Ensure $allMatters is not null and contains elements
if (-not $allMatters) {
    Write-Output "Error: No matters found."
    return $null
}

# Process each Matter object
foreach ($matter in $allMatters) {
    # Check if the matter object has a reference property and that it is not empty
    if ($null -ne $matter['Reference'] -and $matter['Reference'].Length -gt 0) {
        # Extract the part after the first underscore
        $referenceParts = $matter['Reference'] -split "_", 2
        if ($referenceParts.Length -eq 2) {
            $numberPart = $referenceParts[1]  # Extract the number part (after the underscore)
            
            # Initialize the number
            $number = 0

            # Check if the number part is an integer
            if ([int]::TryParse($numberPart, [ref]$number)) {
                # Update the highest number if this number is greater
                if ($number -gt $highestNumber) {
                    $highestNumber = $number
                }
            } else {
                # Debug output if parsing fails
                Write-Output "Failed to parse number part: $($numberPart)"
            }
        } else {
            # Debug output if split fails
            Write-Output "Failed to split reference: $($matter['Reference'])"
        }
    } else {
        # Debug output if reference is null or empty
        Write-Output "Reference is null or empty for matter: $($matter)."
    }
}

# Increment the highest number found
$nextNumber = $highestNumber + 1

# Format the incremented number with leading zeros
$formattedNumber = "{0:D5}" -f $nextNumber  # Ensure 5 digits for the number

# Setup for new matter
$matter = Get-InboundObjectInstance

# Set the matter reference field
$matter.reference = "$($matter.ClientReference)_$($formattedNumber)"

Write-Output "The current highest number is $($highestNumber). The next highest number is $($formattedNumber). $($matter.reference) has been created."
return $matter
```

**Reply by bmward80** [reply]
That's great, thanks for sharing!
